packages/e2e-tests/fixtures/blocks/core__pullquote__main-color__deprecated-4.serialized.html

```diff
@@ -0,0 +1,3 @@
+<!-- wp:pullquote {"className":"has-background is-style-solid-color","style":{"color":{"background":"#48ba90","text":"#1567eb"}}} -->
+<figure class="wp-block-pullquote has-background is-style-solid-color has-text-color has-background" style="background-color:#48ba90;color:#1567eb"><blockquote><p>pullquote</p><cite>before block supports</cite></blockquote></figure>
```

Comment:
Seems like the (text) `color` style is moved from `figure` to `blockquote` during the deprecation:

[Before](https://github.com/WordPress/gutenberg/pull/30951/files#diff-506a54e9a0743b84f3ef2a8ed507e4dfea5186df0e9776386d1852e75374d347):

```html
<figure style="background-color:...">
	<blockquote style="color:...">
		...
	</blockquote>
</figure>
```

After deprecation:

```html
<figure style="background-color:...; color:...">
	<blockquote>
		...
	</blockquote>
</figure>
```

I'm not sure about the impact, but any chance we can keep the HTML unchanged? I think we should be able to use our new `__experimentalSkipSerialization` mechanism to, well skip the "automagic" saving of the `color` attribute, and save it manually instead. cc/ @nosolosw 

(Not really a concern for the deprecation so much as for the block's `save()` method I think, but it's most easily visible here.)

Reply:
> I'm not sure about the impact, but any chance we can keep the HTML unchanged? I think we should be able to use our new __experimentalSkipSerialization mechanism to, well skip the "automagic" saving of the color attribute

So if we use `__experimentalSkipSerialization: true` in `supports.color`, I'm guessing we'd have to do all the work [gutenberg_apply_colors_support()](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php#L66) does, and therefore put back the corresponding color CSS classes (e.g., `has-background-color has-%s-background-color`), and text color style to the blockquote  etc.

Thanks for the sanity check! I'm still getting my head around things. 

> Seems like the (text) color style is moved from figure to blockquote during the deprecation:

I think I agree about re-adding text color style to the block quote, if it avoids specificity/design issues. 

On a more general level however, would we ever accept some level of mutation when "upgrading" the block to use block supports? Especially if the block supports mechanism does not impact the UI in any way when it adds classes/inline styles to the containing block?

Reply:
I believe the normal "ideal scenario" for using the block supports is to have those styles applied to the block's wrapper if at all possible.

In this case there are a few of benefits for applying the block support styles on the block's wrapper element:

- The block's code is a lot simpler, easier to grok and maintain
- Avoids duplicating the generation and application of styles
- Keeps the `__experimentalSkipSerialization` option as a last resort when style must be applied to inner elements.

As a deprecation is already required to migrate the old attributes to the new, it covers the slight change to the HTML avoiding block validation errors. Any CSS specificity concerns, I think would be rather minor, more toward the application of the font color being less specific on the `blockquote` than before.

Overall, my vote would be to continue with the current approach as opposed to using `__experimentalSkipSerialization`.

Reply:
Okay, sounds good. I can't seem to come up with a compelling scenario where moving the text color style would present a huge problem, so let's stick with what we have 👍

Reply:
So, we're changing both the classes and the inline styles for the `color` property from `blockquote` to `figure`. Braindump some use cases in which this could break things:

- is there any element in `figure` that will take color with this change but previously didn't?
- are themes targeting `blockquote.has-color` to do anything (change would be problematic)? Or are they using `figure .has-color` (change would be fine)?


In general, we should be open to changes if it's for the better. We should also check that existing themes don't break. A good litmus tests are the default themes. If we aren't sure about what's the status: either refactor or do behavior changes, not both at the same time.

Reply:
Thanks for the feedback!

> If we aren't sure about what's the status: either refactor or do behavior changes, not both at the same time.

Just checking, by "refactor", do you mean using `__experimentalSkipSerialization` and reattaching the styles to the blockquote?

The behaviour changes, assuming you're referring to the existing background/border logic, must be addressed regardless if we are to update this block to use block supports. So if we do the above refactor we'll have to do both.

It's no problem to refactor, I just wanted to flag it.

I wish there were an MVP path. Ripping off the mixed behaviour band-aid now nevertheless will help folks down the line in my opinion.

It would save a lot of time and energy if we could remove Pullquote's exsiting and inconvenient background/border logic _and_ offer block supports at the same time (which we'd have to do anyway).

> We should also check that existing themes don't break. A good litmus tests are the default themes.

Thanks for this helpful tip.

Despite the direction we take from here, I'll still spend some time testing as many themes as I can.

Given the uncertainty surrounding the consequences of these changes and tendency for cautious conservatism, I think refactoring to accommodate existing HTML structures will ensure there's one less thing to test/worry about. 👍 Happy to take that road, it'll just take a bit more time.

Reply:
I've tested this PR with a range of themes (more theme testing would be beneficial though). 

The majority worked OK as is. 

There were a couple of issues encountered:
- A few themes don't generate or set `has-<color>-border-color` CSS classes so border colors aren't displayed on the frontend. The editor forces the color to an inline style though.
- A couple of themes just needed a `color: inherit` style in the block styles to display correctly
- The TwentyNineteen themes force white text color if the blockquote element itself doesn't have the `has-text-color` class.

| Good | Need `color: inherit` in block styles | Need `has-text-color` on blockquote |
|---|---|---|
| <ul><li>TT1 Blocks</li> <li>TwentyTwentyOne</li> <li>TwentyTwenty</li> <li>TwentyTwenty Blocks</li> <li>Armando</li> <li>Hever</li><li>Q</li><li>Seedlet</li></ul> | <ul><li>Ambitious</li><li>Maywood</li></ul> | <ul><li>Twenty Nineteen</li><li>Twenty Nineteen Blocks</li></ul> |

Themes that don't set the `has-<color>-border-color` styles:
- Ambitious
- Maywood
- TwentyNineteen
- TwentyTwenty

Reply:
Another concern I see is the potential confusion surround the styles.

The "Solid color" doesn't really mean much if you can set (and clear) a custom background. 

Switching between the two is ineffectual (aside from the alignment).

![pullquote-solid-colour](https://user-images.githubusercontent.com/6458278/116028726-296bbb80-a69b-11eb-8dfb-680a2476e59e.gif)

Reply:
Indeed 🤔  Should we consider dropping this UI control, and migrating the attribute to background color (plus maybe a text alignment attribute)?

Reply:
> Should we consider dropping this UI control, and migrating the attribute to background color (plus maybe a text alignment attribute)?

Thanks for the input. 

I agree. 

I'm looking into this right now. It will add to this already large PR, but I don't really see much of an avenue, at least for the future of this PR 🤣 

We can already revert any commits after we've tested them and find that it's over the top.

Reply:
It sounds like a natural evolution of this PR to me. I don't have strong opinions on splitting it into a second PR.

One benefit of adding a `textAlign` attribute and removing the block style now would be keeping the deprecations to a minimum. All the changes driving the need for the deprecation to map the style to attributes would be together and in context.

Reply:
I've removed the styles in https://github.com/WordPress/gutenberg/pull/30951/commits/835caac25a29722063c32549acba011f6045b2ea

I've also added a `textAlign` attribute.

Many themes do target the `.is-style-solid-color` class, however the class name persists in `className` for existing blocks. 

We're not removing any of the existing styles aside from those dealing with text alignment, which the new attribute now takes care of. 

Thanks to everyone for the input. 

Once again, I'm happy to split up this PR if folks things there is a more gentle approach.